bee-network/src/peer/list.rs

```diff
@@ -0,0 +1,469 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use super::{error::Error, meta::PeerState};
+
+use crate::{
+    alias,
+    config::Peer,
+    init::global::max_unknown_peers,
+    swarm::protocols::gossip::io::GossipSender,
+    types::{PeerInfo, PeerRelation},
+};
+
+use libp2p::{Multiaddr, PeerId};
+use tokio::sync::RwLock;
+
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+const REMOTE_PEERS_INITIAL_CAP: usize = 8;
+const LOCAL_ADDRS_INITIAL_CAP: usize = 4;
+
+/// A thread-safe wrapper around a [`PeerList`].
+#[derive(Debug, Clone)]
+pub struct PeerListWrapper(pub Arc<RwLock<PeerList>>);
+
+impl PeerListWrapper {
+    pub fn new(peerlist: PeerList) -> Self {
+        Self(Arc::new(RwLock::new(peerlist)))
+    }
+}
+
+#[derive(Debug)]
+pub struct PeerList {
+    local_id: PeerId,
+    local_addrs: HashSet<Multiaddr>,
+    peers: HashMap<PeerId, (PeerInfo, PeerState)>,
+    banned_peers: HashSet<PeerId>,
+    banned_addrs: HashSet<Multiaddr>,
+}
+
+impl PeerList {
+    pub fn new(local_id: PeerId) -> Self {
+        Self {
+            local_id,
+            local_addrs: HashSet::with_capacity(LOCAL_ADDRS_INITIAL_CAP),
+            peers: HashMap::with_capacity(REMOTE_PEERS_INITIAL_CAP),
+            banned_peers: HashSet::default(),
+            banned_addrs: HashSet::default(),
+        }
+    }
+
+    pub fn from_peers(local_id: PeerId, peers: Vec<Peer>) -> Self {
+        let mut p = HashMap::with_capacity(REMOTE_PEERS_INITIAL_CAP);
+
+        p.extend(peers.into_iter().map(|peer| {
+            (
+                peer.peer_id,
+                (
+                    PeerInfo {
+                        address: peer.multiaddr,
+                        alias: peer.alias.unwrap_or(alias!(peer.peer_id).to_owned()),
+                        relation: PeerRelation::Known,
+                    },
+                    PeerState::new_disconnected(),
+                ),
+            )
+        }));
+
+        Self {
+            local_id,
+            local_addrs: HashSet::with_capacity(LOCAL_ADDRS_INITIAL_CAP),
+            peers: p,
+            banned_peers: HashSet::default(),
+            banned_addrs: HashSet::default(),
+        }
+    }
+
+    pub fn insert_peer(&mut self, peer_id: PeerId, peer_info: PeerInfo) -> Result<(), (PeerId, PeerInfo, Error)> {
+        if self.contains(&peer_id) {
+            return Err((peer_id, peer_info, Error::PeerIsAdded(peer_id)));
+        }
+
+        // Since we already checked that such a `peer_id` is not yet present, the returned value is always `None`.
+        let _ = self.peers.insert(peer_id, (peer_info, PeerState::new_disconnected()));
+
+        Ok(())
+    }
+
+    pub fn insert_local_addr(&mut self, addr: Multiaddr) -> Result<(), (Multiaddr, Error)> {
```

Comment:
maybe using entry would avoid the double access??

Reply:
Certainly can use entry API here :+1:

Reply:
@pvdrz The only problem I have with the entry API in this type of scenario is that it consumes the key no matter what. I think `Entry::Occupied` should also - besides the OccupiedEntry - contain the Key (by value) that was tried to be inserted. What are our thoughts on this?

Reply:
Hmm... yes I think there is no way around adding an extra clone to return the error. I'd left it as it is then